modules/kafka/src/main/scala/com.snowplowanalytics.snowplow.enrich.kafka/Sink.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics.snowplow.enrich.kafka
+
+import cats.Parallel
+import cats.implicits._
+import cats.effect.{Blocker, Concurrent, ConcurrentEffect, ContextShift, Resource, Timer}
+
+import fs2.kafka._
+
+import com.snowplowanalytics.snowplow.enrich.common.fs2.{AttributedByteSink, AttributedData, ByteSink}
+import com.snowplowanalytics.snowplow.enrich.common.fs2.config.io.Output
+
+import java.util.UUID
+
+object Sink {
+
+  def init[F[_]: ConcurrentEffect: ContextShift: Parallel: Timer](
+    blocker: Blocker,
+    output: Output
+  ): Resource[F, ByteSink[F]] =
+    for {
+      sink <- initAttributed(blocker, output)
+    } yield (records: List[Array[Byte]]) => sink(records.map(AttributedData(_, Map.empty)))
+
+  def initAttributed[F[_]: ConcurrentEffect: ContextShift: Parallel: Timer](
+    blocker: Blocker,
+    output: Output
+  ): Resource[F, AttributedByteSink[F]] =
+    output match {
+      case k: Output.Kafka =>
+        mkProducer(blocker, k).map { producer => records =>
+          producer
+            .produce(ProducerRecords(records.map(toProducerRecord(k.topicName, _))))
+            .flatten
+            .void
+        }
+      case o => Resource.eval(Concurrent[F].raiseError(new IllegalArgumentException(s"Output $o is not Kafka")))
+    }
+
+  private def mkProducer[F[_]: ConcurrentEffect: ContextShift](
+    blocker: Blocker,
+    output: Output.Kafka
+  ): Resource[F, KafkaProducer[F, String, Array[Byte]]] = {
+    val producerSettings =
+      ProducerSettings[F, String, Array[Byte]]
+        .withBootstrapServers(output.bootstrapServers)
+        .withProperties(output.producerConf)
+        .withBlocker(blocker)
+
+    KafkaProducer[F].resource(producerSettings)
+  }
+
+  private def toProducerRecord(topicName: String, record: AttributedData[Array[Byte]]): ProducerRecord[String, Array[Byte]] =
+    ProducerRecord(topicName, UUID.randomUUID().toString, record.data)
```

Comment:
We don't need to generate a UUID as Kafka will automatically distribute the records randomly across partitions for us if  no partition and no key is set:
```suggestion
    ProducerRecord(topicName, null, record.data)
```
But we want to be able to specify a partition key in the config. If we specify a key in the `ProducerRecord`, it is automatically used as a partition key:
```
If no partition is provided, and the ProducerRecord has a key, the producer takes the hash of the key modulo the number of partitions. The resulting number from that calculation is the partition that the producer will use.
```
So we want to set the key of `ProducerRecord` to the value of the field set as the `partitionKey` in the config, if defined.

So to answer [your question](https://github.com/snowplow/enrich/pull/659#issuecomment-1201435017) indeed we need to extend `AttributedData` to have both a partition key value and some attributes.

Reply:
thanks Ben, I'll extend AttributedData to carry both

Reply:
addressed in a separate commit